File: AgentLens/Services/Memory/MemoryExtractionParser.swift
import Foundation
import OpenBurnBarCore

// MARK: - Memory Extraction Parser
//
// Strict-JSON parser for the model's extraction output. Pure value type, `Sendable`,
// no I/O. Mirrors `SummaryLLMClient.parseSummaryPayload` (clean-JSON-first, then a
// bounded brace-slice fallback for models that wrap JSON in prose) and hardens it for
// the memory schema: per-candidate validation, confidence clamping, kind mapping,
// length bounds, and a hard candidate ceiling.
//
// The parser is deliberately strict and lossy: a malformed candidate is DROPPED, not
// repaired. An empty / unparseable body yields `[]` (a benign empty — the job still
// succeeds; see the failure taxonomy in `ChatTranscriptExtractor`).

/// One validated candidate the model proposed. `claimedMessageID` is the lookup key
/// the worker resolves against the transcript (PR-D1 must-fix #1) — the parser never
/// treats it as trusted provenance, only as a string the model copied.
struct ExtractedMemoryCandidate: Sendable, Equatable {
    let text: String
    let kind: MemoryKind
    let confidence: Double
    let claimedMessageID: String?
}

enum MemoryExtractionParser {
    /// Maximum characters retained for a single extracted fact body. A durable
    /// memory is a sentence, not a document; this also bounds the work the G7 gate
    /// does per candidate.
    static let maxFactChars = 1_000

    /// Parse `text` into validated candidates, capped at `maxCandidates`.
    static func parse(_ text: String, maxCandidates: Int) -> [ExtractedMemoryCandidate] {
        guard let payload = decodePayload(from: text) else { return [] }
        let cap = max(0, maxCandidates)
        guard cap > 0 else { return [] }

        var results: [ExtractedMemoryCandidate] = []
        for raw in payload.memories {
            if results.count >= cap { break }
            guard let candidate = validate(raw) else { continue }
            results.append(candidate)
        }
        return results
    }

    // MARK: - Decoding

    private static func decodePayload(from text: String) -> RawExtractionPayload? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.isEmpty == false else { return nil }

        if let data = trimmed.data(using: .utf8),
           let decoded = try? JSONDecoder().decode(RawExtractionPayload.self, from: data) { // try?-ok(malformed model JSON falls through)
            return decoded
        }

        // Fallback: slice the outermost JSON object out of a prose-wrapped reply.
        guard let start = trimmed.firstIndex(of: "{"),
              let end = trimmed.lastIndex(of: "}"),
              start < end else {
            return nil
        }
        let candidate = String(trimmed[start ... end])
        guard let data = candidate.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(RawExtractionPayload.self, from: data) // try?-ok(prose-wrapped malformed JSON rejected)
    }

    // MARK: - Validation

    private static func validate(_ raw: RawExtractionPayload.Memory) -> ExtractedMemoryCandidate? {
        guard let text = raw.text else { return nil }
        let body = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard body.isEmpty == false else { return nil }
        let bounded = String(body.prefix(maxFactChars))

        let kind = MemoryKind(rawValue: raw.kind?.lowercased() ?? "") ?? .fact
        let confidence = clampConfidence(raw.confidence)

        let messageID = raw.messageId?.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalizedMessageID = (messageID?.isEmpty == false) ? messageID : nil

        return ExtractedMemoryCandidate(
            text: bounded,
            kind: kind,
            confidence: confidence,
            claimedMessageID: normalizedMessageID
        )
    }

    private static func clampConfidence(_ value: Double?) -> Double {
        guard let value, value.isFinite else { return 0.5 }
        return min(max(value, 0.0), 1.0)
    }

    // MARK: - Raw decodable shapes

    private struct RawExtractionPayload: Decodable {
        struct Memory: Decodable {
            let text: String?
            let kind: String?
            let confidence: Double?
            let messageId: String?

            enum CodingKeys: String, CodingKey {
                case text
                case kind
                case confidence
                case messageId
            }

            init(from decoder: Decoder) throws {
                guard let container = try? decoder.container(keyedBy: CodingKeys.self) else { // try?-ok(non-object memory element is dropped)
                    text = nil
                    kind = nil
                    confidence = nil
                    messageId = nil
                    return
                }
                text = try? container.decode(String.self, forKey: .text) // try?-ok(malformed candidate field is dropped)
                kind = try? container.decode(String.self, forKey: .kind) // try?-ok(malformed optional kind falls back to fact)
                confidence = try? container.decode(Double.self, forKey: .confidence) // try?-ok(malformed optional confidence falls back)
                messageId = try? container.decode(String.self, forKey: .messageId) // try?-ok(malformed optional message id is ignored)
            }
        }

        let memories: [Memory]

        enum CodingKeys: String, CodingKey {
            case memories
        }

        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            memories = (try? container.decode([Memory].self, forKey: .memories)) ?? [] // try?-ok(malformed memories array yields benign empty)
        }
    }
}

File: windows/app/OpenBurnBar.App.MemorySearch/Memory/MemoryExtractionParser.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;

namespace OpenBurnBar.App.MemorySearch.Memory;

// PORTED (faithful) from
//   AgentLens/Services/Memory/MemoryExtractionParser.swift  (parser + ExtractedMemoryCandidate)
//   OpenBurnBarCore/.../Memory/MemoryServing.swift          (enum MemoryKind)
//
// The Swift parser is deliberately strict + lossy: clean-JSON-first, then a bounded
// first-'{' … last-'}' brace-slice fallback for prose-wrapped replies; each candidate field is
// `try?`-decoded so a malformed field is DROPPED/nulled, not a hard error, and a non-object array
// element yields an all-nil (discarded) candidate. Reproduced here with a manual System.Text.Json
// walk because System.Text.Json is strict by default. An empty/unparseable body yields [].

/// <summary>Memory classification. Swift: <c>enum MemoryKind</c> (MemoryServing.swift). Raw-value
/// parity with the lowercase Swift String rawValues.</summary>
public enum MemoryKind
{
    Fact,
    Preference,
    Event,
    Profile,
    Relationship,
    Other,
}

/// <summary>Lowercase raw-value parity + case-insensitive parse. Swift: the <c>String</c>-backed enum.</summary>
public static class MemoryKindExtensions
{
    public static string RawValue(this MemoryKind kind) => kind switch
    {
        MemoryKind.Fact => "fact",
        MemoryKind.Preference => "preference",
        MemoryKind.Event => "event",
        MemoryKind.Profile => "profile",
        MemoryKind.Relationship => "relationship",
        _ => "other",
    };

    /// <summary>Maps a raw value to a kind; unrecognized (or null) → <see cref="MemoryKind.Fact"/>,
    /// matching the parser's <c>MemoryKind(rawValue:) ?? .fact</c>.</summary>
    public static MemoryKind ParseKindOrFact(string? rawLowercased) => rawLowercased switch
    {
        "fact" => MemoryKind.Fact,
        "preference" => MemoryKind.Preference,
        "event" => MemoryKind.Event,
        "profile" => MemoryKind.Profile,
        "relationship" => MemoryKind.Relationship,
        "other" => MemoryKind.Other,
        _ => MemoryKind.Fact,
    };
}

/// <summary>One validated candidate. Swift: <c>struct ExtractedMemoryCandidate</c>.
/// <see cref="ClaimedMessageId"/> is only a lookup key; the worker recomputes provenance.</summary>
public sealed record ExtractedMemoryCandidate(
    string Text,
    MemoryKind Kind,
    double Confidence,
    string? ClaimedMessageId);

/// <summary>Strict-JSON parser for the model's extraction output. Swift: <c>enum MemoryExtractionParser</c>.</summary>
public static class MemoryExtractionParser
{
    /// <summary>Max characters retained for a single fact body. Swift: <c>maxFactChars = 1_000</c>.</summary>
    public const int MaxFactChars = 1_000;

    private readonly record struct RawMemory(string? Text, string? Kind, double? Confidence, string? MessageId);

    /// <summary>Parse <paramref name="text"/> into validated candidates, capped at
    /// <paramref name="maxCandidates"/>. Swift: <c>parse(_:maxCandidates:)</c>.</summary>
    public static List<ExtractedMemoryCandidate> Parse(string text, int maxCandidates)
    {
        ArgumentNullException.ThrowIfNull(text);
        var payload = DecodePayload(text);
        if (payload == null)
        {
            return new List<ExtractedMemoryCandidate>();
        }

        int cap = Math.Max(0, maxCandidates);
        if (cap <= 0)
        {
            return new List<ExtractedMemoryCandidate>();
        }

        var results = new List<ExtractedMemoryCandidate>();
        foreach (var raw in payload)
        {
            if (results.Count >= cap)
            {
                break;
            }

            var candidate = Validate(raw);
            if (candidate != null)
            {
                results.Add(candidate);
            }
        }

        return results;
    }

    private static List<RawMemory>? DecodePayload(string text)
    {
        string trimmed = text.Trim();
        if (trimmed.Length == 0)
        {
            return null;
        }

        var decoded = TryDecode(trimmed);
        if (decoded != null)
        {
            return decoded;
        }

        // Fallback: slice the outermost JSON object out of a prose-wrapped reply.
        int start = trimmed.IndexOf('{');
        int end = trimmed.LastIndexOf('}');
        if (start < 0 || end < 0 || start >= end)
        {
            return null;
        }

        string candidate = trimmed.Substring(start, end - start + 1);
        return TryDecode(candidate);
    }

    /// <summary>
    /// Decodes into a memories list ONLY when the top-level JSON is an object (Swift's
    /// <c>decoder.container(keyedBy:)</c> is non-optional); otherwise returns null so the caller
    /// falls through to the brace slice. A missing/non-array <c>memories</c> → empty list.
    /// </summary>
    private static List<RawMemory>? TryDecode(string json)
    {
        try
        {
            using var document = JsonDocument.Parse(json);
            var root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
            {
                return null;
            }

            var memories = new List<RawMemory>();
            if (root.TryGetProperty("memories", out var memoriesElement)
                && memoriesElement.ValueKind == JsonValueKind.Array)
            {
                foreach (var element in memoriesElement.EnumerateArray())
                {
                    memories.Add(ReadMemory(element));
                }
            }

            return memories;
        }
        catch (JsonException)
        {
            return null;
        }
    }

    private static RawMemory ReadMemory(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            // Non-object array element → all-nil (Swift's keyed-container fallback).
            return new RawMemory(null, null, null, null);
        }

        return new RawMemory(
            ReadString(element, "text"),
            ReadString(element, "kind"),
            ReadDouble(element, "confidence"),
            ReadString(element, "messageId"));
    }

    private static string? ReadString(JsonElement element, string name)
    {
        // Swift `try? decode(String)` only succeeds for a JSON string.
        return element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;
    }

    private static double? ReadDouble(JsonElement element, string name)
    {
        // Swift `try? decode(Double)` only succeeds for a JSON number.
        if (element.TryGetProperty(name, out var value)
            && value.ValueKind == JsonValueKind.Number
            && value.TryGetDouble(out double parsed))
        {
            return parsed;
        }

        return null;
    }

    private static ExtractedMemoryCandidate? Validate(RawMemory raw)
    {
        if (raw.Text == null)
        {
            return null;
        }

        string body = raw.Text.Trim();
        if (body.Length == 0)
        {
            return null;
        }

        string bounded = GraphemePrefix(body, MaxFactChars);
        MemoryKind kind = MemoryKindExtensions.ParseKindOrFact(raw.Kind?.ToLowerInvariant() ?? string.Empty);
        double confidence = ClampConfidence(raw.Confidence);

        string? messageId = raw.MessageId?.Trim();
        string? normalizedMessageId = !string.IsNullOrEmpty(messageId) ? messageId : null;

        return new ExtractedMemoryCandidate(bounded, kind, confidence, normalizedMessageId);
    }

    /// <summary>nil or non-finite → 0.5; else clamped to [0,1]. Swift: <c>clampConfidence(_:)</c>.</summary>
    public static double ClampConfidence(double? value)
    {
        if (value is not double v || !double.IsFinite(v))
        {
            return 0.5;
        }

        return Math.Min(Math.Max(v, 0.0), 1.0);
    }

    private static string GraphemePrefix(string value, int count)
    {
        if (value.Length <= count)
        {
            return value;
        }

        var builder = new StringBuilder();
        var enumerator = StringInfo.GetTextElementEnumerator(value);
        int taken = 0;
        while (taken < count && enumerator.MoveNext())
        {
            builder.Append((string)enumerator.Current);
            taken++;
        }

        return builder.ToString();
    }
}
